docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,136 @@
+# Separate the binding generators into their own crates per target
+
+* Status: proposed
+* Deciders: TBD
+* Date: 2022-01-16
+
+Technical Story: [Issue 299](https://github.com/mozilla/uniffi-rs/issues/299)
+
+outdated early prototype: [PR 997](https://github.com/mozilla/uniffi-rs/pull/997)
+
+## Context and Problem Statement
+All the binding generators currently live in the [`uniffi_bindgen`](../../uniffi_bindgen/src/bindings) crate. This creates the following difficulties:
+
+- All the bindings live in the `uniffi` repository, so the `uniffi` team has to maintain them (or at the very least review changes to them).
+- Any changes to a specific binding generator require a new `uniffi_bindgen` release for it to be released. Even if it doesn't impact any of the other bindings.
+- Some bindings require complex build systems to test, and including those build systems in `uniffi` adds its complexity. For example, any type of `gecko-js` bindings would require the mozilla-central build system to build and test.
+- We currently run all the tests for the bindings in our CI and through `cargo test`. This means that if one binding target gets outdated and fails, or if a developer doesn't have the needed libraries installed for one of the targets, the tests would fail.
+- It's currently difficult to support third-party developers writing bindings for languages the core team does not care about.
+
+This ADR re-evaluates the architecture of how we consume the bindings in hope of improving the developer experience of adding new bindings, testing new and the current bindings and simplifying the ownership of the bindings.
+## Decision Drivers
+
+* Maintainability, we would like to improve the maintainability of our binding generators - this means clear ownership of the bindings.
+* Testability, it should be easy for developers to test the bindings they care about. Without having to navigate and install unfamiliar libraries and build systems.
+* Version compatibility, this refers to the same version of `uniffi` being used to generate both the scaffolding and the bindings. This includes:
+    - How easy is it to accidentally have a version mismatch
+    - How much damage is caused by the the version mismatch
+    - The difficulty of debugging a version mismatch
+* Developer experience, it should be easier to write and maintain a new binding generator than it currently is.
+* Releases, cutting releases for changes in one binding generator shouldn't harm another.
+
+## Considered Options
+
+* **[Option 1] Do nothing**
+
+    This means keeping everything as-is, and deciding that all binding generators (at least for now) should live under the `uniffi_bindgen` crate.
+
+* **[Option 2] Publish a separate crate per binding generator and have users use the  `uniffi_bindgen` Command line interface to shell out to the specific binding generators**
```

Comment:
For this and the next option, how does the `builtin-bindgen` feature work?

Reply:
ideally, it shouldn't change! The `builtin-bindgen` feature should keep using a built-in version of `uniffi-bindgen` for generating the scaffolding code. 

The [`embeded-bindgen`](https://github.com/mozilla/application-services/tree/main/tools/embedded-uniffi-bindgen) we use in app-services, can also stay the same - if a version mismatch happens between the built-in `uniffi_bindgen` in our `embeded_bindgen` and the version of `uniffi_bindgen_kotlin`, error handling logic will log the version mismatch clearly and instruct users to use the right version of `uniffi_bindgen_kotlin`... it's not perfect 😞 but maybe a good start?